File: source/lambda-python/test_cta_client.py
"""Unit tests for CTA-5007-B Python SDK"""
import sys, os, unittest, hmac, hashlib
sys.path.insert(0, os.path.dirname(__file__))
from cta_client import generate_token, parse_ttl, cbor_encode, CWT, CAT, CATU, MATCH

TEST_KEY = 'test-signing-key-for-unit-tests-1234'

class TestParseTTL(unittest.TestCase):
    def test_seconds(self): self.assertEqual(parse_ttl('30s'), 30)
    def test_minutes(self): self.assertEqual(parse_ttl('5m'), 300)
    def test_hours(self): self.assertEqual(parse_ttl('2h'), 7200)
    def test_days(self): self.assertEqual(parse_ttl('1d'), 86400)
    def test_int_passthrough(self): self.assertEqual(parse_ttl(3600), 3600)
    def test_invalid(self): self.assertEqual(parse_ttl('bad'), 7200)

class TestCborEncode(unittest.TestCase):
    def test_int(self): self.assertEqual(cbor_encode(0), b'\x00')
    def test_small_int(self): self.assertEqual(cbor_encode(23), b'\x17')
    def test_one_byte_int(self): self.assertEqual(cbor_encode(24), b'\x18\x18')
    def test_string(self): self.assertEqual(cbor_encode('abc'), b'\x63abc')
    def test_bytes(self): self.assertEqual(cbor_encode(b'\x01\x02'), b'\x42\x01\x02')
    def test_list(self): self.assertEqual(cbor_encode([1, 2]), b'\x82\x01\x02')
    def test_dict(self): self.assertEqual(cbor_encode({1: 5}), b'\xa1\x01\x05')
    def test_none(self): self.assertEqual(cbor_encode(None), b'\xf6')
    def test_bool_true(self): self.assertEqual(cbor_encode(True), b'\xf5')
    def test_bool_false(self): self.assertEqual(cbor_encode(False), b'\xf4')
    def test_negative_int(self): self.assertEqual(cbor_encode(-1), b'\x20')

class TestGenerateToken(unittest.TestCase):
    def setUp(self):
        self.claims = {CWT.ISS: 'test-issuer', CWT.EXP: 1700000000, CWT.IAT: 1699999000}

    def test_returns_bytes(self):
        self.assertIsInstance(generate_token(self.claims, TEST_KEY), bytes)

    def test_cwt_tag(self):
        token = generate_token(self.claims, TEST_KEY)
        self.assertEqual(token[0], 0xd8)
        self.assertEqual(token[1], 0x3d)  # Tag(61) CWT

    def test_cose_mac0_tag(self):
        token = generate_token(self.claims, TEST_KEY)
        self.assertEqual(token[2], 0xd8)
        self.assertEqual(token[3], 0x11)  # Tag(17) COSE_Mac0

    def test_no_cwt_tag(self):
        token = generate_token(self.claims, TEST_KEY, cwt_tag=False)
        self.assertEqual(token[0], 0xd8)
        self.assertEqual(token[1], 0x11)  # Tag(17) directly

    def test_deterministic(self):
        t1 = generate_token(self.claims, TEST_KEY)
        t2 = generate_token(self.claims, TEST_KEY)
        self.assertEqual(t1, t2)

    def test_different_keys(self):
        t1 = generate_token(self.claims, TEST_KEY)
        t2 = generate_token(self.claims, 'different-key')
        self.assertNotEqual(t1, t2)

    def test_hmac_32_bytes(self):
        token = generate_token(self.claims, TEST_KEY)
        self.assertEqual(len(token[-32:]), 32)

    def test_hmac_verifiable(self):
        token = generate_token(self.claims, TEST_KEY)
        protected_bytes = cbor_encode({1: 5})
        payload_bytes = cbor_encode(self.claims)
        mac_structure = cbor_encode(["MAC0", protected_bytes, b'', payload_bytes])
        expected = hmac.new(TEST_KEY.encode(), mac_structure, hashlib.sha256).digest()
        self.assertEqual(token[-32:], expected)

    def test_path_claim(self):
        claims = {CWT.ISS: 'test', CWT.EXP: 1700000000,
                  CAT.CATU: {CATU.PATH: {MATCH.PREFIX: '/video/'}}}
        token = generate_token(claims, TEST_KEY)
        self.assertIn(b'/video/', token)

class TestConstants(unittest.TestCase):
    def test_cwt(self):
        self.assertEqual(CWT.ISS, 1); self.assertEqual(CWT.EXP, 4)
        self.assertEqual(CWT.NBF, 5); self.assertEqual(CWT.IAT, 6); self.assertEqual(CWT.CTI, 7)
    def test_cat(self):
        self.assertEqual(CAT.CATU, 401); self.assertEqual(CAT.CATNIP, 402)
    def test_catu(self):
        self.assertEqual(CATU.PATH, 2); self.assertEqual(MATCH.PREFIX, 1)

if __name__ == '__main__':
    unittest.main()

File: source/lambda-ruby/test_cta_client.rb
require 'minitest/autorun'
require_relative 'cta_client'

TEST_KEY = 'test-signing-key-for-unit-tests-1234'

class TestParseTTL < Minitest::Test
  def test_seconds; assert_equal 30, CTA.parse_ttl('30s'); end
  def test_minutes; assert_equal 300, CTA.parse_ttl('5m'); end
  def test_hours; assert_equal 7200, CTA.parse_ttl('2h'); end
  def test_days; assert_equal 86400, CTA.parse_ttl('1d'); end
  def test_int_passthrough; assert_equal 3600, CTA.parse_ttl(3600); end
  def test_invalid; assert_equal 7200, CTA.parse_ttl('bad'); end
end

class TestCborEncode < Minitest::Test
  def test_zero; assert_equal "\x00".b, CTA.cbor_encode(0); end
  def test_small_int; assert_equal "\x17".b, CTA.cbor_encode(23); end
  def test_one_byte_int; assert_equal "\x18\x18".b, CTA.cbor_encode(24); end
  def test_string; assert_equal "\x63abc".b, CTA.cbor_encode('abc'); end
  def test_bytes; assert_equal "\x42\x01\x02".b, CTA.cbor_encode("\x01\x02".b); end
  def test_list; assert_equal "\x82\x01\x02".b, CTA.cbor_encode([1, 2]); end
  def test_dict; assert_equal "\xa1\x01\x05".b, CTA.cbor_encode({ 1 => 5 }); end
  def test_nil; assert_equal "\xf6".b, CTA.cbor_encode(nil); end
  def test_true; assert_equal "\xf5".b, CTA.cbor_encode(true); end
  def test_false; assert_equal "\xf4".b, CTA.cbor_encode(false); end
  def test_negative; assert_equal "\x20".b, CTA.cbor_encode(-1); end
end

class TestGenerateToken < Minitest::Test
  def setup
    @claims = { CTA::CWT::ISS => 'test-issuer', CTA::CWT::EXP => 1700000000, CTA::CWT::IAT => 1699999000 }
  end

  def test_returns_string
    assert_kind_of String, CTA.generate_token(@claims, TEST_KEY)
  end

  def test_cwt_tag
    token = CTA.generate_token(@claims, TEST_KEY)
    assert_equal 0xd8, token.getbyte(0)
    assert_equal 0x3d, token.getbyte(1)
  end

  def test_cose_mac0_tag
    token = CTA.generate_token(@claims, TEST_KEY)
    assert_equal 0xd8, token.getbyte(2)
    assert_equal 0x11, token.getbyte(3)
  end

  def test_no_cwt_tag
    token = CTA.generate_token(@claims, TEST_KEY, cwt_tag: false)
    assert_equal 0xd8, token.getbyte(0)
    assert_equal 0x11, token.getbyte(1)
  end

  def test_deterministic
    assert_equal CTA.generate_token(@claims, TEST_KEY), CTA.generate_token(@claims, TEST_KEY)
  end

  def test_different_keys
    refute_equal CTA.generate_token(@claims, TEST_KEY), CTA.generate_token(@claims, 'different-key')
  end

  def test_hmac_32_bytes
    token = CTA.generate_token(@claims, TEST_KEY)
    assert_equal 32, token[-32..].bytesize
  end

  def test_hmac_verifiable
    token = CTA.generate_token(@claims, TEST_KEY)
    protected_bytes = CTA.cbor_encode({ 1 => 5 })
    payload_bytes = CTA.cbor_encode(@claims)
    mac_structure = CTA.cbor_encode(["MAC0", protected_bytes, "".b, payload_bytes])
    expected = OpenSSL::HMAC.digest('SHA256', TEST_KEY, mac_structure)
    assert_equal expected, token[-32..]
  end

  def test_path_claim
    claims = { CTA::CWT::ISS => 'test', CTA::CWT::EXP => 1700000000,
               CTA::CAT::CATU => { CTA::CATU::PATH => { CTA::MATCH::PREFIX => '/video/' } } }
    token = CTA.generate_token(claims, TEST_KEY)
    assert_includes token, '/video/'
  end
end

class TestConstants < Minitest::Test
  def test_cwt
    assert_equal 1, CTA::CWT::ISS; assert_equal 4, CTA::CWT::EXP
    assert_equal 5, CTA::CWT::NBF; assert_equal 6, CTA::CWT::IAT; assert_equal 7, CTA::CWT::CTI
  end
  def test_cat
    assert_equal 401, CTA::CAT::CATU; assert_equal 402, CTA::CAT::CATNIP
  end
end
